### PR TITLE
PXP-475: [Wukong CLI] Integrate vault client

### PR DESCRIPTION
**Pull request by mfauzaan**
## Summary

Ticket: [PXP-475](https://mindvalley.atlassian.net/browse/PXP-475)

## What's Changed

- [x]  implement ability to Login via Okta credential
- [x]  add ability to store the Vault token. 
- [x]  add ability to introspect the token to check the validity. 
- [x]  ability to read secrets
- [x]  ability to update a secret
 

[PXP-475]: https://mindvalley.atlassian.net/browse/PXP-475?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/76#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by jk-gan — CHANGES_REQUESTED** [review]
There is no need for a new vault command. This ticket only need a vault client that can do `login` and etc (refer to the Jira ticket). 

This will allow us to connect to Vault in the future when we are implementing `wukong dev config-synthesizer` and `wukong dev config-secrets-edit` commands.

File: Cargo.toml
Comment (Fadhil) [inline Cargo.toml:72]: Ahaa! I can use this to reimplement my solution with recursion. Nice one.

File: src/services/vault/mod.rs
Comment (onimsha) [inline src/services/vault/mod.rs:90]: @jk-gan @mfauzaan since each dev already have to perform `wukong init` and use there email for the Okta auth, should we also imply that the same email will be use for Vault authentication as well ? It seems redundant to me that they will need to re-enter their email again. Given that the Wukong CLI is not supporting profiles yet we can assume that each Wukong config only has 1 profile ?

Comment (mfauzaan) [reply src/services/vault/mod.rs:90]: That make sense, let me refactor a bit to use the existing profile

File: src/services/vault/client.rs
Comment (Fadhil) [inline src/services/vault/client.rs:63]: Should instead build the client inside `new` and then reuse it with `self.client`?

Comment (mfauzaan) [reply src/services/vault/client.rs:63]: Thats a good catch, yeah I shoud've done that

File: src/services/vault/mod.rs
Comment (mfauzaan) [reply src/services/vault/mod.rs:205]: Yes, i've tested out and both invalid token and expired token returns same code ie 403

Invalid Token:
Screenshot 2023-03-21 at 12 56 06 PM

Expired Token:
Screenshot 2023-03-21 at 2 03 14 PM

File: src/services/vault/mod.rs
Comment (mfauzaan) [reply src/services/vault/mod.rs:96]: Yes, Tested out and login failure falls to 400 as well 

Screenshot 2023-03-21 at 2 06 29 PM

File: src/services/vault/mod.rs
Comment (jk-gan) [inline src/services/vault/mod.rs:177]: same as above

File: src/services/vault/mod.rs
Comment (jk-gan) [inline src/services/vault/mod.rs:198]: for the error message, I think we should use the `CLIError` to do the output, which is in `main.rs` file

File: src/services/vault/mod.rs
Comment (jk-gan) [inline src/services/vault/mod.rs:203]: same as above

File: src/services/vault/mod.rs
Comment (jk-gan) [inline src/services/vault/mod.rs:206]: same as above

File: src/services/vault/mod.rs
Comment (jk-gan) [inline src/services/vault/mod.rs:214]: same as above

**Review by jk-gan — CHANGES_REQUESTED** [review]

File: src/services/vault/client.rs
Comment (onimsha) [inline src/services/vault/client.rs:311]: I thought we're using `POST` ? Or am i missing something here ?

Comment (mfauzaan) [reply src/services/vault/client.rs:311]: Yeah, missed that update. Updating now

File: src/services/vault/mod.rs
Comment (jk-gan) [reply src/services/vault/mod.rs:195]: My thinking is that since this is just a helper function, we don't make too much side effect like showing the error on to the screen. We could just return the error value and the caller can decide when and how to display the error.

For example
```rust
// in `commands/dev/config_xxx.rs`
let vault = Vault::new();
match vault.get_secrets {
    Ok(x) => { ... },
    Err(error) => { // decide whether to display error here or not }
}
```

Comment (mfauzaan) [reply src/services/vault/mod.rs:195]: Meaning i dont redirect to login function if user is not logged in?

Comment (jk-gan) [reply src/services/vault/mod.rs:195]: Unless you think the error should be displayed here, and we want to do continue other actions instead of end the program. Then you should use `eprintln!()` to display on the `stderr` channel

Comment (mfauzaan) [reply src/services/vault/mod.rs:196]: Hmm, that works

So we'll keep it for now with eprintln!, as we discussed since this is CLI logic we might move this whole thing away

Comment (jk-gan) [reply src/services/vault/mod.rs:196]: Ok that works too. We will figure this out on the next ticket when we're implementing the command and actually using these helper functions. Let's explore together then. For now, just change the error output to `eprintln`

Comment (jk-gan) [reply src/services/vault/mod.rs:196]: It is a good chance to learn about good API/function design.

**Review by jk-gan — APPROVED** [review]
LGTM

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/76#pullrequestreview-1353889214" title="Approved">jk-gan</a>*, *<a href="https://github.com/mindvalley/wukong-cli/pull/76#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
